### PR TITLE
Improve detach of volume

### PR DESCRIPTION
**Pull request by mazzy89**
Towards https://github.com/giantswarm/giantswarm/issues/6555#issuecomment-514163631

This implements a backoff for the detach of the volumes.

This is inspired by the upstream logic implemented in the cloud controller manager.

It is still not tested so in order to test I'd have to recreate likely the condition in that PM.

### CODE REVIEW

File: service/controller/legacy/v29/ebs/ebs.go [hunk 1]
```diff
@@ -128,6 +128,61 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
...
+				}
+			}
+
+			return nil
+		}
+
+		b := backoff.NewExponential(30*time.Second, 5*time.Second)
+		n := backoff.NewNotifier(e.logger, context.Background())
+
+		err := backoff.RetryNotify(o, b, n)
```
Comment (xh3b4sd) [inline service/controller/legacy/v29/ebs/ebs.go]: That is the wrong approach because this blocks the resource and thus all other reconciliation, which is what we want to prevent. All the checking on the volumes can be done, but we should not backoff anymore within resources. When the detaching is not yet done, the resource should be canceled. It is then executed again anyway on the next reconciliation loop after the regular resync period.

Comment (mazzy89) [reply service/controller/legacy/v29/ebs/ebs.go]: So what you recommend here it is simply to remove the backoff func and leave the logic like it is. Did I understand correctly?

File: service/controller/legacy/v29/ebs/ebs.go [hunk 2]
```diff
@@ -128,6 +128,61 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
...
+
+			v, err := e.client.DescribeVolumes(i)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			if len(v.Volumes) == 0 {
+				// Fall through.
+				e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("volume %#q not found", volumeID))
+				return nil
```
Comment (xh3b4sd) [inline service/controller/legacy/v29/ebs/ebs.go]: We should return `executionFailedError`.

File: service/controller/legacy/v29/ebs/ebs.go [hunk 3]
```diff
@@ -128,6 +128,61 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
...
+			}
+
+			if len(v.Volumes) > 1 {
+				// Fall through.
+				e.logger.LogCtx(ctx, "level", "debug", "message", "multiple volumes found")
+				return nil
```
Comment (xh3b4sd) [inline service/controller/legacy/v29/ebs/ebs.go]: We should return `executionFailedError`.

File: service/controller/legacy/v29/ebs/ebs.go [hunk 4]
```diff
@@ -128,6 +128,61 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
...
+			if len(d.Attachments) > 1 {
+				// Shouldn't happen; log so we know if it is
+				e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found multiple attachment for volume %#q", volumeID))
+			}
+
+			for _, a := range d.Attachments {
+				if a.State != nil {
+					attachmentStatus := *a.State
+					if attachmentStatus != "detached" {
+						return microerror.New(fmt.Sprintf("the volume %#q is still attached", volumeID))
```
Comment (xh3b4sd) [inline service/controller/legacy/v29/ebs/ebs.go]: This should be a typed error which we can match against in the resource. When we then match against the returned error, we cancel the resource.

File: service/controller/legacy/v29/ebs/ebs.go (hunk 1 above)
Comment (xh3b4sd) [reply service/controller/legacy/v29/ebs/ebs.go]: As discussed we want to remove the backoff magic here.

**Comment by mazzy89** [comment]
Applied the changes as discussed today

File: service/controller/legacy/v29/ebs/ebs.go [hunk 5]
```diff
@@ -128,6 +128,43 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
...
+				aws.String(volumeID),
+			},
+		}
+
+		v, err := e.client.DescribeVolumes(i)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+
+		if len(v.Volumes) == 0 {
```
Comment (xh3b4sd) [inline service/controller/legacy/v29/ebs/ebs.go]: Why did you change the conditions? There are maybe less or more than 1 volume and we want to return an error in both cases I think.

File: service/controller/legacy/v29/ebs/ebs.go [hunk 6]
```diff
@@ -128,6 +128,43 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
...
+			// Fall through.
+			e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("volume %#q not found", volumeID))
+			return microerror.Maskf(executionFailedError, "no volume found")
```
Comment (xh3b4sd) [inline service/controller/legacy/v29/ebs/ebs.go]: Let's do either logging or returning an error. Both is an anti-pattern IMO. So let's just drop the log above and this is fine.

File: service/controller/legacy/v29/ebs/ebs.go [hunk 7]
```diff
@@ -128,6 +128,43 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
...
+		if len(d.Attachments) > 1 {
+			// Shouldn't happen; log so we know if it is
+			e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("found multiple attachment for volume %#q", volumeID))
+		}
+
+		for _, a := range d.Attachments {
+			if a.State != nil {
+				attachmentStatus := *a.State
+				if attachmentStatus != "detached" {
+					return microerror.Maskf(volumeAttachedError, "volume %#q still attached", volumeID)
```
Comment (xh3b4sd) [inline service/controller/legacy/v29/ebs/ebs.go]: We still need the error matching in the calling resource. Otherwise the resource returns an error, gets retried by the `retryresource` and then causes the whole loop to fail. As discussed, we want to match for the error we expect as long as the volume is still attached and cancel the resource gracefully.

File: service/controller/legacy/v29/ebs/error.go [hunk 8]
```diff
@@ -5,6 +5,10 @@ import (
 	"github.com/giantswarm/microerror"
 )
 
+var executionFailedError = &microerror.Error{
```
Comment (xh3b4sd) [inline service/controller/legacy/v29/ebs/error.go]: Let's keep that in sync like in https://github.com/giantswarm/aws-operator/pull/1796.

```suggestion
// executionFailedError is an error type for situations where Resource execution
// cannot continue and must always fall back to operatorkit.
//
// This error should never be matched against and therefore there is no matcher
// implement. For further information see:
//
//     https://github.com/giantswarm/fmt/blob/master/go/errors.md#matching-errors
//
var executionFailedError = &microerror.Error{
```

File: service/controller/legacy/v29/ebs/error.go [hunk 9]
```diff
@@ -40,3 +44,7 @@ func IsVolumeNotFound(err error) bool {
 
 	return false
 }
+
+var volumeAttachedError = &microerror.Error{
```
Comment (xh3b4sd) [inline service/controller/legacy/v29/ebs/error.go]: Let's please always have custom errors and their matchers together as described in https://github.com/giantswarm/fmt/blob/master/go/errors.md. As mentioned in the other comment above, we need the matching in the calling resource anyway.

File: service/controller/legacy/v29/ebs/ebs.go (hunk 5 above)
Comment (mazzy89) [reply service/controller/legacy/v29/ebs/ebs.go]: Because we spoke during our call and you said clearly that if we do the lookup with the volumeID then the only possible result is one volume. Not more than one. So i removed that condition.

File: service/controller/legacy/v29/ebs/ebs.go (hunk 7 above)
Comment (mazzy89) [reply service/controller/legacy/v29/ebs/ebs.go]: I'm not familiar with this kind of implementation. I will take a look in the other part of the codebase to see how it is implemented

[the remark this answers is not in the document]
Comment (xh3b4sd) [reply service/controller/legacy/v29/ebs/ebs.go]: I don't think so because we want to match against that error and then cancel the resource gracefully.

File: service/controller/legacy/v29/ebs/ebs.go [hunk 10]
```diff
@@ -128,6 +128,43 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
 		if err != nil {
 			return microerror.Mask(err)
 		}
+	}
+
```
[the remark this answers is not in the document]
Comment (mazzy89) [reply service/controller/legacy/v29/ebs/ebs.go]: No definitely also in the clusterapi

File: service/controller/legacy/v29/ebs/ebs.go [hunk 11]
```diff
@@ -128,6 +128,43 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
...
+			if a.State != nil {
+				attachmentStatus := *a.State
+				if attachmentStatus != "detached" {
+					return microerror.Maskf(volumeAttachedError, "volume %#q still attached", volumeID)
+				}
+			} else {
+				// Shouldn't happen; log so we know if it is
+				e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("nil attachment state for volume %#q", volumeID))
+			}
+		}
```
[the remark this answers is not in the document]
Comment (mazzy89) [reply service/controller/legacy/v29/ebs/ebs.go]: Modified

File: service/controller/legacy/v29/ebs/ebs.go (hunk 5 above)
Comment (xh3b4sd) [reply service/controller/legacy/v29/ebs/ebs.go]: We said there should only be one volume and if there is none or more than one we should return an error. I made comments for that. The conditions have been correct as you implemented them, only the error handling needed improvements.

File: service/controller/legacy/v29/ebs/ebs.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (xh3b4sd) [reply service/controller/legacy/v29/ebs/ebs.go]: We want to get the implementation right once and then move it over.

File: service/controller/legacy/v29/ebs/ebs.go (hunk 7 above)
Comment (xh3b4sd) [reply service/controller/legacy/v29/ebs/ebs.go]: You can see this in https://github.com/giantswarm/aws-operator/blob/master/service/controller/clusterapi/v29/resource/tccpsubnet/create.go where `addSubnetsToContext` returns an error that is matched against in `EnsureCreated` where we then cancel the reconciliation.

File: service/controller/legacy/v29/ebs/ebs.go [hunk 12]
```diff
@@ -128,6 +128,43 @@ func (e *EBS) DetachVolume(ctx context.Context, volumeID string, attachment Volu
...
+
+		if len(v.Volumes) == 0 {
+			// Fall through.
+			e.logger.LogCtx(ctx, "level", "debug", "message", fmt.Sprintf("volume %#q not found", volumeID))
+			return microerror.Maskf(executionFailedError, "no volume found")
+		}
+
+		d := v.Volumes[0]
+		if len(d.Attachments) > 1 {
+			// Shouldn't happen; log so we know if it is
```
[the remark this answers is not in the document]
Comment (mazzy89) [reply service/controller/legacy/v29/ebs/ebs.go]: This is an approach borrowed from the upstream implementation. If the volume has multiple attachments does not mean that there is an error. More that we need to be aware of this and log. But at the end since we have a range below we'll loop through it

File: service/controller/legacy/v29/ebs/ebs.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (kopiczko) [reply service/controller/legacy/v29/ebs/ebs.go]: I see.

File: service/controller/legacy/v29/resource/ebsvolume/error.go [hunk 13]
```diff
@@ -12,3 +12,12 @@ var invalidConfigError = &microerror.Error{
 func IsInvalidConfig(err error) bool {
 	return microerror.Cause(err) == invalidConfigError
 }
+
+var volumeAttachedError = &microerror.Error{
+	Kind: "volumeAttachedError",
+}
+
+// isVolumeAttached asserts volumeAttachedError.
+func isVolumeAttached(err error) bool {
```
Comment (xh3b4sd) [inline service/controller/legacy/v29/resource/ebsvolume/error.go]: That does not make sense. When you test this code you will see that the resource will never be cancelled because the matcher matches against the `volumeAttachedError` of the `ebs` service package, which will never match. We do not need the error and matcher here in the resource. We only need to use the matcher of the `ebs` service in the resource. Also not that matcher should always be public as described by `fmt`.

File: service/controller/legacy/v29/resource/ebsvolume/delete.go [hunk 14]
```diff
@@ -59,7 +59,11 @@ func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
 				shutdown := true
 				wait := false
 				err := ebsService.DetachVolume(ctx, vol.VolumeID, a, force, shutdown, wait)
-				if err != nil {
+				if isVolumeAttached(err) {
```
Comment (xh3b4sd) [inline service/controller/legacy/v29/resource/ebsvolume/delete.go]:
```suggestion
				if ebs.IsVolumeAttached(err) {
```

File: service/controller/legacy/v29/resource/ebsvolume/delete.go [hunk 15]
```diff
@@ -73,7 +77,11 @@ func (r *Resource) EnsureDeleted(ctx context.Context, obj interface{}) error {
 				shutdown := false
 				wait := false
 				err := ebsService.DetachVolume(ctx, vol.VolumeID, a, force, shutdown, wait)
-				if err != nil {
+				if isVolumeAttached(err) {
```
Comment (xh3b4sd) [inline service/controller/legacy/v29/resource/ebsvolume/delete.go]:
```suggestion
				if ebs.IsVolumeAttached(err) {
```

**Comment by mazzy89** [comment]
@xh3b4sd now i understand what you mean. I've fixed. thank you

**Comment by mazzy89** [comment]
Thanks a lot to everyone for this great PR feedbacks. learn a lot around the operator logic. ;-) 

going to port in clusterapi and then i'll give a test to check everything

**Comment by mazzy89** [comment]
All good here. ported and i gave a quick run. all good. merging.
